text/mavlink2_1.md

```diff
@@ -0,0 +1,62 @@
+  * Start date: 2022-10-12
+  * Contributors: Andrew Tridgell
+
+# Summary
+
+Rough proposal for MAVLink 2.1
+
+# Motivation
+
+Key features wanted:
+
+ * support for more than 256 system IDs
+ * support for checking CRCs in routers that don't have all xml
+ * better target_system/target_component support
+ * maximum compatibility with existing mavlink2 systems
+
+# Detailed Design
+
+The design takes advantage of the incompat_flags and compat_flags
+fields in the MAVLink2 header.
+
+New incompat_flags flags would be:
+
+ * MAVLINK_IFLAG_SYSID16 : system ID in header is 16 bits wide
+ * MAVLINK_IFLAG_TARGETTED : header has target_system and target_component appended
+ * MAVLINK_IFLAG_SPLIT_CRC : 16 bit crc is as described below
+
+New compat_flags flags would be:
+
+ * MAVLINK_FLAG_21_CAPABLE : sender is capable of handling MAVLink 2.1
+
+Note that for the SYSID16 and TARGETTED flags the sender should only
+use the flag if is it needed. So if the systemID in the message is
+<=255 then the flag should not be set and the existing 8 bit field
+should be used. For the TARGETTED bit this should only be used if the
+target systemID is >255 (needing 16 bits) or if the message that needs
+targetting doesn't have a target_system or target_component field.
+
+Following these rules maximises compatibility with existing MAVLink2
+capable systems.
+
+Note that when SYSID16 flag is set then the target_system would be 16
+bits when TARGETTED is set.
+
+For messages with existing target_system/target_component fields, if
+the TARGETTED bit is set then the existing fields are ignored by the
+recipient. The sender should set the old fields to zero.
+
+The 16 bit size of the expanded systemID could perhaps be 32 bits,
+allowing for an IPv4 address to be used, which could be useful for
+many use cases. The target audience is drone light shows which often
+use WiFi.
+
+## Split CRC
```

Comment:
For terminology, for "structural seed" do you mean the [CRC_EXTRA](https://mavlink.io/en/guide/serialization.html#crc_extra) - i.e. the CRC created by running the calculation over the name of the message and the and types of the field in over-the-wire order.

So is the proposal that you have run the whole CRC for the message contents (16 bits)  then apply the CRC_EXTRA somehow to just the last 8 bits? Or something else?

Reply:
yes, CRC_EXTRA is the structural seed, it aims to checksum the structure
unfortunately my proposal means you'd need to calculate the CRC twice

Reply:
OK. I still don't understand the proposed algorithm, but perhaps others will.

Right now I think it is it is
- to send: run CRC algorithm over each field in order to calculate the CRC of the message, then accumulate the CRC_EXTRA over the top of that. Result a 2 byte CRC 
- on receive: run CRC over incoming data, then apply your CRC Extra. Numbers should match. 
- Problem is that you can't separate CRC and CRC_EXTRA - you don't know if a failure is message being corrupt or incompatible and you can't work that out without having the CRC_EXTRA matching the library.

What I think you are suggesting is that you can separate these by applying the CRC_EXTRA just on the final 8 bits of the CRC, but I don't understand the encode/decode algorithm.

Reply:
i think this works:
 - calculate crc16 over message bytes as transmitted on wire. call this 16 bit value c1
 - b1 = c1&0xFF
 - continue crc16 over the CRC_EXTRA bytes call the resulting crc c2
 - b2 = c2 >> 8
 - c3 = b1<<8 | b2

crc used in header is c3

Reply:
Thanks. Hopefully someone who knows what they are doing can sanity check that :-)

Reply:
Trying to confirm I "grok" this ...

Presumably this is all assuming a >16 bit data type the code is working on (hence the need to do `c1&0xFF`) so when you send c3 you just take the bottom 2 bytes?

My understanding of this is:
- A router on receiving could calculate b1 and compare to the "b1 byte" of `c3` it got in the message. If this matched this would indicate a pass of the CRC check so the router could forward the message; if not it would know the message was corrupt and could drop it.
- At a final end points the receiver can compare the first byte to check for bit corruption, and the second byte to check for either bit corruption or message mismatch.

If so, I can see this will algorithm works. Is there a loss of data compared to the current situation, and is it a concern? I mean, you now have a single byte of data that derives from the CRC_EXTRA calculation. That means that if you have more than 256 messages defined won't they have the same number - and does this increase the chance of deciding that the sender/receiver match when they don't?